### PR TITLE
Improvements to Inject.

### PR DESCRIPTION
**Pull request by sellout**
- move it back to cats-core (near `Coproduct`)
- make `inj`/`prj` natural transformations
- add `apply`/`unapply` (lowered `inj`/`prj`)
- define `Free.roll`
- move `injectRoll` (née `Inject.inject`) and `match_` to `Free`
- reintroduce "null identity" test (written by @edmundnoble)

### CODE REVIEW

**Comment by sellout** [comment]
These are largely changes needed to ease Quasar’s migration.

File: tests/src/test/scala/cats/tests/InjectTests.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package cats
+
+// import cats.arrow.FunctionK
```
Comment (adelbertc) [inline tests/src/test/scala/cats/tests/InjectTests.scala]: woops

File: core/src/main/scala/cats/Inject.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Inject type class as described in "Data types a la carte" (Swierstra 2008).
+ *
+ * @see [[http://www.staff.science.uu.nl/~swier004/publications/2008-jfp.pdf]]
+ */
+sealed abstract class Inject[F[_], G[_]] {
+  def inj: FunctionK[F, G]
+
+  def prj: FunctionK[G, λ[α => Option[F[α]]]]
```
Comment (adelbertc) [inline core/src/main/scala/cats/Inject.scala]: Should we call these `inject` and `project` ?

File: tests/src/test/scala/cats/tests/InjectTests.scala (hunk 1 above)
Comment (sellout) [inline tests/src/test/scala/cats/tests/InjectTests.scala]: Also, I apparently broke tut. Ok, fixes in the morning 😅

File: core/src/main/scala/cats/Inject.scala (hunk 2 above)
Comment (sellout) [inline core/src/main/scala/cats/Inject.scala]: I’m fine with either. The current ones match the paper and Scalaz. Also, Matryoshka defines `project`, so I don’t mind the lack of collision 😁

**Comment by notxcain** [comment]
@sellout did you see https://github.com/typelevel/cats/issues/1505 in the wild? If so, how do you workaround?

**Comment by sellout** [comment]
@notxcain Sorry, I haven’t seen that.

File: core/src/main/scala/cats/Inject.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package cats
+
+import cats.arrow.FunctionK
+import cats.data.Coproduct
+import cats.syntax.either._
```
Comment (kailuowang) [inline core/src/main/scala/cats/Inject.scala]: unused import here breaking the build

Comment (sellout) [inline core/src/main/scala/cats/Inject.scala]: Thanks. I have a false memory of seeing a green build, so this would have sat for a while before I noticed.

**Comment by kailuowang** [comment]
Sorry my bad. The either syntax is used in scala 2.11 and 2.10 withi this [`toOption`](https://github.com/sellout/cats/blob/d9566a215f3cf8846d79fd58b87dab19ac954a58/core/src/main/scala/cats/Inject.scala#L33). It's unused in Scala 2.12 because `Either` is right biased and `toOption` runs on the right projection. So now the build on 2.10 and 2.11 is broken. A fix off top of my head is to write `_.run.swap.right.toOption` instead. There may be a better way to support the different syntax in different Scala? cc @adelbertc

**Comment by sellout** [comment]
@kailuowang Yeah – I brought this up on the gitter channel. @paulp mentioned ghik/silencer, but that doesn’t support 2.10. I’ll try the `swap` approach. Thanks.

**Comment by kailuowang** [comment]
👍  LGTM, since this is a breaking change, we might want to wait for a third maintainer sign-off.

**Comment by kailuowang** [comment]
merging with three sign-offs

**Pull request merged**